sidecar/tlv.go

```diff
@@ -0,0 +1,386 @@
+// data to be encoded in the tlv format.
+func DeserializeTicket(r io.Reader) (*Ticket, error) {
+	var (
+		ticket                     = &Ticket{}
+		version, state             uint8
+		offerBytes, recipientBytes []byte
+		orderBytes, executionBytes []byte
+	)
+
+	tlvStream, err := tlv.NewStream(
+		tlv.MakeStaticRecord(idType, &ticket.ID, 8, EBytes8, DBytes8),
+		tlv.MakePrimitiveRecord(versionType, &version),
+		tlv.MakePrimitiveRecord(stateType, &state),
+		tlv.MakePrimitiveRecord(offerType, &offerBytes),
+		tlv.MakePrimitiveRecord(recipientType, &recipientBytes),
+		tlv.MakePrimitiveRecord(orderType, &orderBytes),
+		tlv.MakePrimitiveRecord(executionType, &executionBytes),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedTypes, err := tlvStream.DecodeWithParsedTypes(r)
+	if err != nil {
+		return nil, err
+	}
+
+	ticket.Version = Version(version)
+	ticket.State = State(state)
+
+	if t, ok := parsedTypes[offerType]; ok && t == nil {
+		ticket.Offer, err = deserializeOffer(offerBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if t, ok := parsedTypes[recipientType]; ok && t == nil {
+		recipient, err := deserializeRecipient(recipientBytes)
+		if err != nil {
+			return nil, err
+		}
+		ticket.Recipient = &recipient
+	}
+
+	if t, ok := parsedTypes[orderType]; ok && t == nil {
+		order, err := deserializeOrder(orderBytes)
+		if err != nil {
+			return nil, err
+		}
+		ticket.Order = &order
+	}
+
+	if t, ok := parsedTypes[executionType]; ok && t == nil {
+		execution, err := deserializeExecution(executionBytes)
+		if err != nil {
+			return nil, err
+		}
+		ticket.Execution = &execution
+	}
+
+	return ticket, nil
+}
+
+// serializeOffer serializes the given offer to a byte slice using the tlv
+// format.
+func serializeOffer(o Offer) ([]byte, error) {
+	capacity := uint64(o.Capacity)
+	pushAmt := uint64(o.PushAmt)
+
+	tlvRecords := []tlv.Record{
+		tlv.MakePrimitiveRecord(capacityType, &capacity),
+		tlv.MakePrimitiveRecord(pushAmtType, &pushAmt),
+	}
+
+	if o.SignPubKey != nil {
+		tlvRecords = append(tlvRecords, tlv.MakePrimitiveRecord(
+			signPubKeyType, &o.SignPubKey,
+		))
+	}
+
+	if o.SigOfferDigest != nil {
+		tlvRecords = append(tlvRecords, tlv.MakeStaticRecord(
+			sigOfferDigestType, &o.SigOfferDigest, 64, ESig, DSig,
+		))
+	}
+
+	return encodeBytes(tlvRecords...)
+}
+
+// deserializeOffer deserializes an order from the given byte slice, expecting
+// the data to be encoded in the tlv format.
+func deserializeOffer(offerBytes []byte) (Offer, error) {
+	var (
+		o                 = Offer{}
+		capacity, pushAmt uint64
+	)
+
+	if err := decodeBytes(
+		offerBytes,
+		tlv.MakePrimitiveRecord(capacityType, &capacity),
+		tlv.MakePrimitiveRecord(pushAmtType, &pushAmt),
+		tlv.MakePrimitiveRecord(signPubKeyType, &o.SignPubKey),
+		tlv.MakeStaticRecord(
+			sigOfferDigestType, &o.SigOfferDigest, 64, ESig, DSig,
+		),
+	); err != nil {
+		return o, err
+	}
+
+	o.Capacity = btcutil.Amount(capacity)
```

Comment:
probably doesn't matter, but check type was parsed before setting it for consistency?

Reply:
never mind, later I see we decode directly into the structs, which is kinda nice way to shorten the code.